### PR TITLE
Explicit-comms: dataframe shuffle

### PR DESCRIPTION
**Pull request by madsbk**
This PR implements Dataframe shuffle using explicit-comms.

cc. @rjzamora

### CODE REVIEW

File: dask_cuda/explicit_comms/dataframe_shuffle.py [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+    of workers.
+
+    See Also
+    --------
+    set_index
+    set_partition
+    shuffle_disk
+    """
+
+    if isinstance(index, str) or (
```
Comment (rjzamora) [inline dask_cuda/explicit_comms/dataframe_shuffle.py:227]: It looks like this is the only code path with test coverage, and only for the `str` case.  We should add a test for a list of `str` column names, and maybe for a collection-based index.

File: dask_cuda/explicit_comms/dataframe_shuffle.py [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+    containing the output partition number of every row.  This function
+    produces a new dataframe where every row is in the proper partition.  It
+    accomplishes this by splitting each input partition into several pieces,
+    and then concatenating pieces from different input partitions into output
+    partitions.
+
+    Note that the `column` input may correspond to a list of columns (rather
+    than just a single column name).  In this case, the `shuffle_group` and
+    `shuffle_group_2` functions will use hashing to map each row to an output
+    partition. This approach may require the same rows to be hashed multiple
```
Comment (rjzamora) [inline dask_cuda/explicit_comms/dataframe_shuffle.py:176]: "multiple times" comment only applies to staged shuffling

File: dask_cuda/explicit_comms/dataframe_shuffle.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+    eps = {trim_map[i]: s["eps"][trim_map[i]] for i in workers if i != s["rank"]}
+
+    df = df_concat(df_parts)
+
+    return await rearrange_by_column(len(workers), rank, eps, df, column)
+
+
+def dataframe_rearrange_by_column(df, column):
+    """Order divisions of DataFrame so that all values within column(s) align
+
```
Comment (rjzamora) [inline dask_cuda/explicit_comms/dataframe_shuffle.py:163]: Perhaps its best to state that this is an "explicit-comms" version of `rearrange_by_column_tasks`?

File: dask_cuda/explicit_comms/dataframe_shuffle.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+    bins.extend(await asyncio.gather(*futures))
+
+
+async def exchange_and_concat_bins(rank, eps, bins):
+    ret = [bins[rank]]
+    await asyncio.gather(recv_bins(eps, ret), send_bins(eps, bins))
+    return concat([df for df in ret if df is not None])
+
+
+def concat(df_list):
```
Comment (rjzamora) [inline dask_cuda/explicit_comms/dataframe_shuffle.py:55]: Can we use [`_concat`](https://github.com/dask/dask/blob/a6615bd175bec61ea62e7d5187ea7d6e2da2dee7/dask/dataframe/core.py#L83)?

**Review by rjzamora — COMMENTED** [review]
Thanks for putting this together @madsbk !

I made a bunch of minor comments, but the overall approach looks good to me :)

**Comment by madsbk** [comment]
Thanks for the review @rjzamora, I have simplified much of code and documentation. Please give it another go when you have time.

File: dask_cuda/explicit_comms/dataframe_shuffle.py (hunk 4 above)
Comment (madsbk) [reply dask_cuda/explicit_comms/dataframe_shuffle.py]: Done

File: dask_cuda/explicit_comms/dataframe_merge.py [hunk 5]
```diff
@@ -228,66 +122,36 @@ def df_concat(df_parts):
...
 
 
-def dataframe_merge(left, right, on=None, left_on=None, right_on=None, how="inner"):
-    """Merge two Dask DataFrames
+def dataframe_merge(left, right, on=None, left_on=None, right_on=None):
+    """Merge two DataFrames using explicit-comms.
 
-    This will merge the two datasets, either on the indices, a certain column
-    in each dataset or the index in one dataset and the column in another.
+    This is an explicit-comms version of Dask's Dataframe.merge() that
```
Comment (rjzamora) [inline dask_cuda/explicit_comms/dataframe_merge.py:128]: No need to address in this Pr,  but "left" is pretty common and shouldn't be difficult to support (although, I may be missing something)

**Review by rjzamora — APPROVED** [review]
This is looking good @madsbk - I really like the simplifications you made :)

**Comment by madsbk** [comment]
Thanks @rjzamora!

**Comment by madsbk** [comment]
rerun tests

**Pull request merged**